Implemented sender compendium comparison tool / obx mapping tool

thetaurean opened the pull request:
This adds a command line utility for comparing a sender compendium to a loaded observation mapping table.

Test Steps:
1. Load the observation mapping table
```./gradlew primeCLI --args "lookuptables create -e local -i ./obs-mapping.csv -a -n obs"```
2. Run the comparison
```./gradlew primeCLI --args "lookuptables compare-mapping -e local -i ./compend.csv -n obs"```

Files:
- [compend.csv](https://github.com/CDCgov/prime-reportstream/files/13242032/compend.csv)
- [obs-mapping.csv](https://github.com/CDCgov/prime-reportstream/files/13242033/obs-mapping.csv)

## Changes
- A new CLI tool for comparing an observation mapping tool
- One private function for saving to csv moved to a general utility class
- Two tests added

## Checklist

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] Added tests?

## Linked Issues
- Fixes #11935

JFU-NAVA-PBC reviewed (commented):
Looks good generally, the command line has reasonable args parsing and error messages.

and the comparison report looks good:

![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/f2cbc21c-1cc6-4b30-81fe-72d2a65c22fd)

JFU-NAVA-PBC commented:
there might be some over validation of input file?

if I run compare with the compend.csv where there are 3 columns:

I got error as below and command stop proceeding:

`Supplied compendium is missing column: mapped?`

===================
test code,test description,coding system
97099-6,Influenza virus A and B and SARS-CoV-2 (COVID-19) Ag panel - Upper respiratory specimen by Rapid immunoassay,LOINC
47457-7,Influenza virus A H8 Ab [Titer] in Serum,LOINC
123456, LDT Flu Test,LOCAL
49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,LOINC
49526-7,NOMATCH,LOINC
49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,NOMATCH
===================

I have to patch and populate the input as below to make the compare command happy and echo out result:

=====================
test code,test description,coding system,mapped?
97099-6,Influenza virus A and B and SARS-CoV-2 (COVID-19) Ag panel - Upper respiratory specimen by Rapid immunoassay,LOINC,X
47457-7,Influenza virus A H8 Ab [Titer] in Serum,LOINC,X
123456, LDT Flu Test,LOCAL,X
49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,LOINC,X
49526-7,NOMATCH,LOINC,X
49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,NOMATCH,X
=====================

May be the input validation on mapped? column should be lifted?

JFU-NAVA-PBC commented:
Also, did not see test coverage ;-)

JFU-NAVA-PBC commented:
otherwise LGTM

JFU-NAVA-PBC commented:
Like the human review in the table upload flow:

![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/92100dbe-158e-412e-865d-2e3d3b602bbd)

JFU-NAVA-PBC commented on prime-router/src/main/kotlin/cli/LookupTableCommands.kt:
```diff
@@ -14,7 +14,6 @@ import com.github.ajalt.clikt.parameters.types.int
 import com.github.difflib.text.DiffRow
 import com.github.difflib.text.DiffRowGenerator
 import com.github.doyaaaaaken.kotlincsv.dsl.csvReader
-import com.github.doyaaaaaken.kotlincsv.dsl.csvWriter
```
Good move of 3rd party csv processor into Report Stream util class so that all the Commands class does not directly import from the 3rd party packages...

JFU-NAVA-PBC reviewed (approved):
left comments and approve

thetaurean commented:
> there might be some over validation of input file?
> 
> if I run compare with the compend.csv where there are 3 columns:
> 
> I got error as below and command stop proceeding:
> 
> `Supplied compendium is missing column: mapped?`
> 
> ++++++++ test code,test description,coding system 97099-6,Influenza virus A and B and SARS-CoV-2 (COVID-19) Ag panel - Upper respiratory specimen by Rapid immunoassay,LOINC 47457-7,Influenza virus A H8 Ab [Titer] in Serum,LOINC 123456, LDT Flu Test,LOCAL 49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,LOINC 49526-7,NOMATCH,LOINC 49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,NOMATCH ++++++++
> 
> I have to patch and populate the input as below to make the compare command happy and echo out result:
> 
> +++++++++ test code,test description,coding system,mapped? 97099-6,Influenza virus A and B and SARS-CoV-2 (COVID-19) Ag panel - Upper respiratory specimen by Rapid immunoassay,LOINC,X 47457-7,Influenza virus A H8 Ab [Titer] in Serum,LOINC,X 123456, LDT Flu Test,LOCAL,X 49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,LOINC,X 49526-7,NOMATCH,LOINC,X 49526-7,Influenza virus A H5 RNA [Presence] in Isolate by NAA with probe detection,NOMATCH,X +++++++++
> 
> May be the input validation on mapped? column should be lifted?

Fixed! Sorry that wasn't supposed to be there. Nice catch 👍

thetaurean commented:
Changes pushed, should be ready for final checks now, but a quick re-check wouldn't hurt @JFU-NAVA-PBC
- Added tests
- Minor cleanup: added comment and moved 1 function for easier testing
- Fixed validation bug

Thanks!

JFU-NAVA-PBC reviewed (approved):
Nice, checked out and passed all tests

arnejduranovic reviewed (approved):
Looks great!! Have one comment for naming suggestions but not a big deal. Tested locally with a larger dataset and found no issues. Tried making values null, changing codes, etc.
